.github/workflows/ci.yml

```diff
@@ -379,12 +381,63 @@ jobs:
           composer config allow-plugins.silverstripe/recipe-plugin true
           composer config allow-plugins.silverstripe/vendor-plugin true
 
-          # Useful to see generated composer.json when diagnosing new bugs
-          cat composer.json
-
           # matrix.composer_args sometimes includes `--prefer-lowest` which is only supported by `composer update`, not `composer install`
           # Modules do not have composer.lock files, so `composer update` is the same speed as `composer install`
-          composer update --no-interaction --no-progress ${{ matrix.composer_args }}
+          # `|| :` prevents an exit code on a failed composer update from halting the workflow
+          composer update --no-interaction --no-progress ${{ matrix.composer_args }} 2> __update_attempt.txt || :
+
+          if ! [[ $(cat __update_attempt.txt) =~ Problem ]]; then
+            # Succesfully ran composer update and installed everything
+            rm __update_attempt.txt
+            # Useful to see generated composer.json when diagnosing new bugs
+            cat composer.json
+          elif ! [[ $(cat __update_attempt.txt) =~ 'Root composer.json requires silverstripe/installer' ]] || \
+            [[ $INPUTS_COMPOSER_REQUIRE_EXTRA =~ silverstripe/installer ]] || \
+            [[ $MATRIX_COMPOSER_REQUIRE_EXTRA =~ silverstripe/installer ]] || \
+            ! [[ "${{ matrix.installer_version }}" =~ ^([1-9])\.([0-9]+)\.x\-dev$ ]]
+          then
+            # Failed to run composer update and will not attempt requiring different versions of silverstripe/installer
+            cat composer.json
+            cat __update_attempt.txt
+            exit 2
+          else
+            # gha-generate-matrix will not always provide a compatibile version of silverstripe/installer due to
+            # limitations of knowing what versions of silverstripe/installer the pushed code is compatible with
+            # In this scenario we will usually end up with the latest minor x.dev version of silverstripe installer
+            # If it did initially work on composer update, try some earlier versions of silverstripe/installer
+            rm __update_attempt.txt
+            echo "Could not run composer update with version of silverstripe/installer, attempting earlier versions"
+            MAJOR=${BASH_REMATCH[1]}
```

Comment:
Using `BASH_REMATCH` so many lines after the regex seems a bit fragile, especially since getting these values isn't the primary purpose of that regex.

Reply:
Why is it fragile? It's the last regex that ran.  That regex has groups in it so it is one of the purposes of the regex

Reply:
It's _currently_ the last regex that ran. But if we need to amend this logic and end up adding another regex in here, will we notice that it's used down here to get these values?

Reply:
ok will de-fragilize